src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,571 @@
+
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+    function _max(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x >= y ? x : y;
+    }
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _add(x, _sub(y, 1)) / y;
+    }
+
+    // --- Administration ---
+    /**
+        @notice update vow or end addresses.
+        @dev msg.sender must be authorized.
+        @param what name of what we are updating bytes32("vow"|"end")
+        @param data address we are setting it to
+    */
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+
+        if (what == "vow") vow = data;
+        else if (what == "end") end = EndLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    /**
+        @notice update tau value for D3M ilk.
+        @dev msg.sender must be authorized.
+        @param ilk  bytes32 of the D3M ilk to be updated
+        @param what bytes32("tau") or it will revert
+        @param data number of second to wait after caging a pool to write of debt
+    */
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "tau" ) {
+            ilks[ilk].tau = data;
+        } else revert("DssDirectDepositHub/file-unrecognized-param");
+
+        emit File(ilk, what, data);
+    }
+
+    /**
+        @notice update plan or pool addresses for D3M ilk.
+        @dev msg.sender must be authorized.
+        @param ilk  bytes32 of the D3M ilk to be updated
+        @param what bytes32("pool"|"plan") or it will revert
+        @param data address we are setting it to
+    */
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "pool") ilks[ilk].pool = D3MPoolLike(data);
+        else if (what == "plan") ilks[ilk].plan = D3MPlanLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, D3MPoolLike pool, uint256 amount) internal {
```

Comment:
Consider moving `internal` functions after `external` ones and grouping them accordingly (see dss-vest).